File: versioncontrol/src/main/java/ir/ayantech/versioncontrol/api/VCReasonModel.java
package ir.ayantech.versioncontrol.api;

import android.util.Log;

import com.google.gson.Gson;

import java.io.IOException;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.util.concurrent.TimeoutException;

import ir.ayantech.versioncontrol.model.VCResponseModel;
import retrofit2.Response;

/**
 * Created by shadoWalker on 5/9/17.
 */

public abstract class VCReasonModel {

    private static final String NO_INTERNET = "دستگاه شما به اینترنت متصل نیست. لطفا بعد از بررسی دوباره تلاش نمایید.";
    private static final String NO_HOST = "ارتباط با سرور برقرار نشد. لطفا دوباره تلاش نمایید.";

    protected void handleError(VersionControlAPI offerAPI, Throwable t, VCResponseStatus status) {
        boolean callback = true;
        boolean canTry = true;
        String message;
        if (t instanceof UnknownHostException)
            message = NO_INTERNET;
        else if (t instanceof TimeoutException)
            message = NO_HOST;
        else if (t instanceof SocketTimeoutException)
            message = NO_HOST;
        else if (t instanceof SocketException) {
            canTry = false;
            message = "";
            callback = false;
        } else if (t instanceof IOException) {
            canTry = false;
            message = "";
            callback = true;
        } else
            message = NO_HOST;
        if (callback)
            status.onFail(offerAPI, message, canTry);
    }

    public boolean isCodeOk(int code) {
        if (code == 200)
            return true;
        return false;
    }

    public <T extends VCResponseModel> T convertJsonStringToObject(String json, Class<T> object) {
        return new Gson().fromJson(json, object);
    }

    public <T extends VCResponseModel> T handleResponse(Response<T> response, Class<T> tClass) throws IOException {
        if (isCodeOk(response.code())) {
            return response.body();
        } else {
            String error = response.errorBody().string();
            return convertJsonStringToObject(error, tClass);
        }
    }
}

File: versioncontrol/src/main/java/ir/ayantech/versioncontrol/api/VCReasonModel.kt
package ir.ayantech.versioncontrol.api

import com.google.gson.Gson
import ir.ayantech.versioncontrol.model.VCResponseModel
import retrofit2.Response
import java.io.IOException
import java.net.SocketException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.concurrent.TimeoutException

abstract class VCReasonModel {

    companion object {
        private const val NO_INTERNET = "دستگاه شما به اینترنت متصل نیست. لطفا بعد از بررسی دوباره تلاش نمایید."
        private const val NO_HOST = "ارتباط با سرور برقرار نشد. لطفا دوباره تلاش نمایید."
    }

    protected open fun handleError(
        offerAPI: VersionControlAPI<*, *>,
        t: Throwable?,
        status: VCResponseStatus
    ) {
        var callback = true
        var canTry = true
        val message: String
        when (t) {
            is UnknownHostException -> message = NO_INTERNET
            is TimeoutException, is SocketTimeoutException -> message = NO_HOST
            is SocketException -> {
                canTry = false
                message = ""
                callback = false
            }
            is IOException -> {
                canTry = false
                message = ""
                callback = true
            }
            else -> message = NO_HOST
        }
        if (callback) {
            status.onFail(offerAPI, message, canTry)
        }
    }

    open fun isCodeOk(code: Int): Boolean {
        return code == 200
    }

    open fun <T : VCResponseModel> convertJsonStringToObject(json: String?, objectClass: Class<T>): T {
        return Gson().fromJson(json, objectClass)
    }

    @Throws(IOException::class)
    open fun <T : VCResponseModel> handleResponse(response: Response<T>, tClass: Class<T>): T? {
        return if (isCodeOk(response.code())) {
            response.body()
        } else {
            val error = response.errorBody()?.string()
            convertJsonStringToObject(error, tClass)
        }
    }
}
